FlightTaskAuto: get landing altitude from mission

kibidev opened the pull request:
### Solved Problem
RTL and mission landing assumes that ground level is the same as home, even though the correct ground level is given as altitude in the mission item.

Fixes #21086

### Solution
~~Override _evaluateDistanceToGround in FlightTaskAuto, and use mission altitude (in landing mission item only).~~
FlightTaskAuto::_prepareLandSetpoints() will use altitude base one (in prioritised order):
1. Range sensor measurement
2. Altitude from mission item or RP
3. Home altitude*

*In addition, if range sensor or mission/RP altitude is not available, a new parameter `MPC_LAND_MODE` selects if home altitude should be assumed, or if velocity should be set to `MPC_Z_V_AUTO_DN` or `MPC_LAND_SPEED`.

### Alternatives
Probably a few other places to correct for this, at least in _evaluateDistanceToGround in FlightTask, or in FlightTaskAuto.

### Test coverage
Briefly tested in simulator for mission landing, RP landing and commanded landing.
Atm poor coverage, much TODO.

### Context
N/A

kibidev committed:
FlightTaskAuto: get landing altitude from mission

If altitude setpoint is availible in auto landing, it will be used to
calculate distance to ground.

ThomasRigi commented:
Thanks @kibidev for the PR. I unfortunately don't have time to look into it properly, as I leave for a 5-month vacation in a week. At first glance your changes look promising and well contained, but we need to make sure that there are no uncovered cases where `_sub_triplet_setpoint.get().current.alt` would be badly set when Landing. I reckon Mission landings should be fine (as long as the planned mission follows mavlink specs of https://mavlink.io/en/messages/common.html#MAV_CMD_NAV_LAND), but I don't know if for instance a simple Land command (in the sense of a "simple" mode change, for example as a failsafe action or when clicking on the land button in e.g. QGC) would also work as desired. And maybe I'm forgetting something, we'd have to check the code for all `SETPOINT_TYPE_LAND` occurrences at least to make sure.

I would have a preference to set the land altitude to 0m AMSL (or -500m to catch all weird places on earth ;)) if nothing is known about the ground altitude, and risk a hard landing in case of no lidar rather than a slow landing due to a random dependency on home altitude, but I know that's a debatable opinion, and I wouldn't want this discussion to block a clear improvement on RTL and mission landings. 

@KonradRudin would you have time to check this ? Or know someone to tag to look into it ? ;)

KonradRudin commented:
Hey, 
I am not very familiar with the flight tasks but @MaEtUgR could be a review candidate for this.
As for the implementation, as @ThomasRigi already mentioned, we would need to check on the other modes setting the landing. On the landing mode itsel,f it calls the function set_land_item from the mission_block module. There the landing altitude is hardcoded to 0 https://github.com/aviant-tech/PX4-Autopilot/blob/landing_altitude_from_mission_in_auto/src/modules/navigator/mission_block.cpp#L861. We would need to change that at least to NaN (or in a future iteration actually use the value given from the mavlink message). 
Also the same change would be needed for the precision landing. Here https://github.com/aviant-tech/PX4-Autopilot/blob/landing_altitude_from_mission_in_auto/src/modules/navigator/precland.cpp#L300 we should set the altitude of the ground altitude as well (the information should be in the _target_pose variable).

kibidev commented:
Thanks for input. I'll look into the issues, but probably not this week.
I think the most important feedback this early is simply if anyone sees a conceptional problem regarding using _evaluateDistanceToGround override to user RP and mission landing altitude to calculate the distance to the ground.
It is what makes most sens to me, but I'm not 100% happy with "almost copy" the implementation from the super class. Also find the current way that distance sensor is used first in _evaluateDistanceToGround, then again in FlightTaskAuto::_prepareLandSetpoints to evauate soft landing to be a source to confusion, and more so if I override _evaluateDistanceToGround. But I guess that is perhaps a different issue that can be done in a later refactoring anyway.

junwoo091400 commented:
@kibidev Thanks for this PR, any update?

kibidev commented:
> @kibidev Thanks for this PR, any update?

Hi @junwoo091400, I have not have time to work more on this, and probably will not have time in the near future. Anyone should feel free to take this further (or solve the issue using an other approach) if the want, otherwise I'll continue whenever I have time.

kibidev commented:
@MaEtUgR
I have moved most of the changes to _prepareLandSetpoints as you suggest.
It is still a work in progress.

I not set altitude setpoint to NAN rather than 0 in MissionBlock::set_land_item(). I also had to change FlightTaskAuto::_evaluateTriplets() slightly to accept NAN as altitude when landing.

I have also added a new parameter to be considered, to be able to chose behavior when the ground level is not known. It covers the use cases I can imaging on the fly. If you fly VLOS in a flat field, you can assume home altitude when commanding land. If you have lots of battery and want to touch down lightly, you can chose to use MPC_LAND_SPEED, and if battery is an issue, and its better to have a hard controlled landing that risk to run out of battery, you can choose to use MPC_Z_V_AUTO_DN.
